### PR TITLE
Implement Pearce SCC algorithm

### PR DESCRIPTION
**Pull request by oxidase**
# Issue

For #1392 Pearce SCC algorithm is similar to Tarjan SCC, but requires slightly less memory, but in worst case it uses the same amount of memory as Tarjans SCC. Time performance is nearly the same. 

Don't know if it is useful to use the Pearce SCC algorithm instead of the current one.

## Tasklist
 - [ ] review
 - [ ] adjust for comments

## Requirements / Relations
 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: include/extractor/pearce_scc.hpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+{
+
+template <typename GraphT> class PearceSCC
+{
+    std::vector<NodeID> components_index;
+    std::vector<std::size_t> component_size_vector;
+    std::shared_ptr<const GraphT> m_graph;
+
+  public:
+    PearceSCC(std::shared_ptr<const GraphT> graph)
```
Comment (daniel-j-h) [inline include/extractor/pearce_scc.hpp]: I don't understand why we need to take shared ownership over a graph here.

A `const GraphT&` should be good enough? The callsite can unwrap if needs be.

File: include/extractor/pearce_scc.hpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    {
+    }
+
+    void Run()
+    {
+        TIMER_START(SCC_RUN);
+        const NodeID max_node_id = m_graph->GetNumberOfNodes();
+
+        // recursion stack contains local variables: v, (w, end(w)), root
+        std::stack<std::tuple<NodeID, typename GraphT::EdgeRange, bool>> recursion_stack;
```
Comment (daniel-j-h) [inline include/extractor/pearce_scc.hpp]: Maybe change from `tuple` and `get<n>` below to dedicated named local struct.

File: unit_tests/util/scc.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                       {3, 2},
+                       {3, 7},
+                       {4, 0},
+                       {4, 5},
+                       {5, 6},
+                       {6, 5},
+                       {7, 3},
+                       {7, 6}}};
+    // osrm::extractor::TarjanSCC<Graph> scc(std::make_shared<const Graph>(graph));
+    osrm::extractor::PearceSCC<Graph> scc(std::make_shared<const Graph>(graph));
```
Comment (daniel-j-h) [inline unit_tests/util/scc.cpp]: No need for shared ownership.

**Comment by daniel-j-h** [comment]
Can you benchmark this? Would be cool to see memory and runtime on the planet.

File: include/extractor/pearce_scc.hpp (hunk 1 above)
Comment (oxidase) [inline include/extractor/pearce_scc.hpp]: it is not needed, just copied constructor arguments from `TarjanSCC`

**Comment by daniel-j-h** [comment]
We're currently seeing the partitioner taking roughly 40 GB for the planet.
And the Tarjan run before using above 50 GB on its own.

Might make sense to try this one out.

**Comment by oxidase** [comment]
Just checked: the current implementation has no memory benefit over the original implementation :cold_sweat:. The problem is in `StackFrame` has a flag `root`  that is 32 bits and `edges` range that is 64 bits that can be reduced. The paper promises only memory reduction from `v(2+5w)` to `v(1+3w)` where `w` is 32 bits. Will try to fit to `v(1+3w)` estimation

**Comment by daniel-j-h** [comment]
@oxidase even though it would be a nice to have feature, I would not prioritize this.

**Comment by TheMarex** [comment]
Didn't pay off. :crying_cat_face:

**Pull request closed without merging**